APM UI changes for serverless services / AWS lambda

gbamparop opened the pull request:
Closes https://github.com/elastic/kibana/issues/106552

Requires https://github.com/elastic/apm-server/issues/6576 to properly work with metric documents.

## Summary
Some APM agents now support AWS lambda instrumentations. In this context some of the data is not available while some additional, serverless-specific information is available that is very valuable for APM use cases. This PR contains UI changes to reflect these aspects for serverless services.

With AWS lambda currently all kind of metric collection are disabled at the agents. Thus, metrics for system, JVM, transaction breakdown are not available.

https://user-images.githubusercontent.com/5831975/151546017-8b4f3604-06a5-430c-b6c0-20f3f04484ee.mov

### Changes

* Added meta-information for AWS lambda services in the service-icons panel
![image](https://user-images.githubusercontent.com/5831975/151544983-12fee08b-8e6d-4e60-83a9-e44bf6d87537.png)

* Added `cloud.service.name` and `cloud.region` (before only availability zone was shown) as additional info being shown in cloud metainfo section 
![image](https://user-images.githubusercontent.com/5831975/150371974-144aa15b-f537-4d8e-bffa-babd9f5bf11a.png)
* Replaced transaction breakdown chart with cold start rate chart for serverless services
  * In service overview
![image](https://user-images.githubusercontent.com/866830/141441209-563628cc-79dc-4f0e-839c-f6b3c96cc66b.png)

  * In transactions overview
![image](https://user-images.githubusercontent.com/5831975/150334151-cf86e6ca-0a48-4fba-9a5b-3de882c11e79.png)

  * In transaction details
![image](https://user-images.githubusercontent.com/866830/141441244-4f44c1cc-5ff7-4eec-b2a7-6ea9cb470d84.png)
* Added `cold start` badge to waterfall transaction items (when they represent cold start requests)
![image](https://user-images.githubusercontent.com/866830/141488916-0c96f191-a201-4ce9-b797-c621fe97c83a.png)
* Added `cold start` badge to transaction flyout (when it represents cold start requests)
![image](https://user-images.githubusercontent.com/5831975/150342220-1a6b7bc1-c474-4c8d-a415-991242e9fef7.png)

* Hidden metrics-related views (for which metrics are not available)
  * Instances latency distribution chart and instance table
  * Metrics and JVM tab 

## Testing
### Synthtrace
Run the following scenario with [synthrace CLI](https://github.com/gbamparop/kibana/blob/aws-lambda/packages/elastic-apm-synthtrace/README.md#cli)

[aws_lambda_with_multiple_functions.ts.zip](https://github.com/elastic/kibana/files/7958488/aws_lambda_with_multiple_functions.ts.zip)

### Cloud
@gbamparop can help to set this up.

1) Create a deployment using the latest snapshot (e.g. 8.1.0-SNAPSHOT)
2) Add lambda extension and instrument a lambda function by following [this guide](https://www.elastic.co/guide/en/apm/agent/nodejs/current/lambda.html)
3) Invoke the lambda function to ingest data in APM
4) Connect local kibana instance to the cloud

## Todo
- [ ]  Docs: Additions for charts and [transaction](https://github.com/elastic/kibana/blob/main/docs/apm/transactions.asciidoc) view changes etc (https://github.com/elastic/observability-docs/issues/1490) (pending screenshots with real data)

Commit:
Adapted service UI for AWS lambda / serverless services

Commit:
Refactor lambda UI for main branch merge

AlexanderWert committed:
Update x-pack/plugins/apm/public/components/shared/charts/transaction_coldstart_rate_chart/index.tsx

Co-authored-by: Casper Hübertz <casper@formgeist.com>

gbamparop committed:
Fix linting errors

gbamparop committed:
Fix API tests

gbamparop committed:
Add unit tests for isServerlessAgent function

gbamparop committed:
Add story for cold start badge

gbamparop committed:
Add unit tests for service icons and icon details

gbamparop committed:
Add aws_lambda checks to isMetricsTabHidden and isJVMsTabHidden unit tests

gbamparop committed:
Add API test for coldstart_rate chart

gbamparop committed:
Use termQuery helper for filters

gbamparop committed:
Remove noHits from cold start rate chart response

gbamparop committed:
Change service icon API tests to use synthrace and test for serverless property

gbamparop committed:
Change service details API tests to use synthrace and add test for serverless

gbamparop commented:
@formgeist shall we add the `coldstart` badge to the transaction flyout as well? Also, if there's a requirement to have a `beta` badge where do you think it's the best place to add it?

gbamparop committed:
Fix sync_badge path

gbamparop committed:
Revert generate_data file

gbamparop committed:
Move sync_badge tests to subfolder

gbamparop committed:
Add e2e tests for cold start rate chart

formgeist commented:
> @formgeist shall we add the `coldstart` badge to the transaction flyout as well?

Yes, I think it makes sense similar to the `async` badge in this example;

<img width="668" alt="CleanShot 2022-01-19 at 19 32 42@2x" src="https://user-images.githubusercontent.com/4104278/150192637-a4256cf1-638a-461c-b1b2-2e3968e89abc.png">

> Also, if there's a requirement to have a beta badge for this piece of work where do you think it's the best place to add it?

I think the only place where I think it has the most impact is when the user clicks on the Lambda service icon and looks at the content. So adding a `EuiBetaBadge` ([EUI docs](https://eui.elastic.co/#/display/badge#beta-badge-type)) which will show a tooltip on hover.

![Slice 1](https://user-images.githubusercontent.com/4104278/150193632-0196c05d-e349-43b1-8806-0e4a21ebe1c3.png)

cc @alex-fedotyev

gbamparop committed:
Change app metrics to transactions in API tests

gbamparop committed:
Add cold start badge to transaction flyout

gbamparop commented:
@formgeist thanks, I'll get those added

gbamparop committed:
Add beta badge to cloud details in a lambda context

gbamparop committed:
Change faas.coldstart type to boolean

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

gbamparop commented:
@elasticmachine merge upstream

alex-fedotyev commented:
@gbamparop - do you know why docker icon is displayed next to lambda icon on some of the screenshots?
Is that discrepancy in your test data, or something we need to have logic for to better prioritize?

cc @AlexanderWert

AlexanderWert commented:
> do you know why docker icon is displayed next to lambda icon on some of the screenshots?
Is that discrepancy in your test data, or something we need to have logic for to better prioritize?

I'd guess it's because the data comes from Synthtrace which in addition to the Lambda fields also generated some docker fields. Shouldn't happen with real data. 

@gbamparop correct me if I'm wrong.

formgeist commented:
On a related note, has anyone opened an issue for adding a Lambda service to our fleet of Opbeans test data for continued testing?

gbamparop commented:
> > do you know why docker icon is displayed next to lambda icon on some of the screenshots?
> > Is that discrepancy in your test data, or something we need to have logic for to better prioritize?
> 
> I'd guess it's because the data comes from Synthtrace which in addition to the Lambda fields also generated some docker fields. Shouldn't happen with real data.
> 
> @gbamparop correct me if I'm wrong.

gbamparop commented:
> On a related note, has anyone opened an issue for adding a Lambda service to our fleet of Opbeans test data for continued testing?

I haven't, do you know how we could do that?

dgieselaar commented:
@gbamparop we can file an issue in https://github.com/elastic/apm-integration-testing

gbamparop commented:
> @gbamparop we can file an issue in https://github.com/elastic/apm-integration-testing

Thanks, I've created https://github.com/elastic/apm-integration-testing/issues/1405

dgieselaar commented on x-pack/plugins/apm/server/routes/transactions/route.ts:
```diff
@@ -364,10 +365,64 @@ const transactionChartsErrorRateRoute = createApmServerRoute({
+    }),
+    query: t.intersection([
+      t.type({ transactionType: t.string }),
+      t.partial({ transactionName: t.string }),
```
Personally I'd prefer to separate routes, one where transaction name is required, and one where it is not allowed. We've seen a couple of bugs where having optional parameters led to us forgetting to add them in the appropriate places.

gbamparop replied:
Added another route with required `transactionName`

dgieselaar commented on x-pack/test/apm_api_integration/tests/cold_start/generate_data.ts:
```diff
@@ -0,0 +1,56 @@
+
+  const traceEvents = timerange(start, end)
+    .interval('1m')
+    .rate(rate)
```
does it make sense to generate the same rate/duration for both? if we want to make sure data is properly filtered we should probably use different values for cold and warm starts.

gbamparop replied:
Good point, added different rates for the transactions

dgieselaar commented on x-pack/test/apm_api_integration/tests/cold_start/cold_start.spec.ts:
```diff
@@ -0,0 +1,174 @@
+          );
+        });
+
+        it('has same average value for both periods', () => {
```
ideally we index different values (and test for them as well)

gbamparop replied:
Added different values and now testing for all of them

dgieselaar commented on x-pack/plugins/apm/server/routes/services/get_service_metadata_details.ts:
```diff
@@ -104,12 +114,30 @@ export async function getServiceMetadataDetails({
             size: 10,
           },
         },
+        regions: {
```
I think at this point we should revise our fetching strategy. Each added terms aggregation is another iteration of all the matching documents, and this adds three terms aggregations. Let's figure out what fields are available on what documents and see if we can build up a query that returns a single document that has them all.

gbamparop replied:
As discussed, this request is made only when a service metadata icon is clicked and it can be addressed in another PR. Raised https://github.com/elastic/kibana/issues/123957

dgieselaar commented on x-pack/plugins/apm/server/lib/transaction_groups/get_coldstart_rate.ts:
```diff
@@ -0,0 +1,181 @@
+
+  const filter = [
+    ...termQuery(SERVICE_NAME, serviceName),
+    ...termsQuery(FAAS_COLDSTART, 'true', 'false'),
```
should this be an exists query?

gbamparop replied:
Changed it to exists filter

dgieselaar commented on x-pack/plugins/apm/server/lib/helpers/transaction_coldstart_rate.ts:
```diff
@@ -0,0 +1,67 @@
+export function calculateTransactionColdstartRate(
+  coldstartStatesResponse: AggregationResultOf<ColdstartAggregation, {}>
+) {
+  const coldstartStates = Object.fromEntries(
+    coldstartStatesResponse.buckets.map(({ key, doc_count: count }) => [
+      key === 1 ? 'true' : 'false',
+      count,
+    ])
+  );
+
+  const coldstarts = coldstartStates.true ?? 0;
+  const warmstarts = coldstartStates.false ?? 0;
+
+  return coldstarts / (coldstarts + warmstarts);
```
There's probably a more straight-forward way to express this. What about using something like lodash's `partition`?

```ts
const [ warmStarts, coldStarts ] = partition(coldstartStatesResponse, ( bucket ) => bucket.key === 1);

return coldStarts.length / (coldStarts.length - warmStarts.length);
```

dgieselaar commented on x-pack/plugins/apm/server/routes/transactions/route.ts:
```diff
@@ -364,10 +365,64 @@ const transactionChartsErrorRateRoute = createApmServerRoute({
+    ]),
+  }),
+  options: { tags: ['access:apm'] },
+  handler: async (resources) => {
```
Can you add an explicit return type here? VS Code should give you a refactor action that will add it.

gbamparop committed:
Add return type to cold start chart route

gbamparop committed:
Update cold start rate tests to use different transaction rates

gbamparop committed:
Change FAAS_COLDSTART to 'exists' filter

gbamparop committed:
Fix ts error

gbamparop committed:
Create another route for cold start rate chart with required transaction name

gbamparop committed:
Add API tests for coldstart_rate_by_transaction_name endpoint

gbamparop committed:
Add support for multiple lambda functions in a single service

gbamparop commented:
@formgeist I've added support for multiple lambda functions in a single service and added a screenshot of the lambda service icon details panel in the description. Are there any changes that would you like there?

dgieselaar commented on x-pack/plugins/apm/public/components/shared/charts/transaction_coldstart_rate_chart/index.tsx:
```diff
@@ -0,0 +1,184 @@
+    comparisonEnabled,
+  });
+
+  const endpoint = transactionName
+    ? ('GET /internal/apm/services/{serviceName}/transactions/charts/coldstart_rate_by_transaction_name' as const)
+    : ('GET /internal/apm/services/{serviceName}/transactions/charts/coldstart_rate' as const);
```
I was expecting the data fetching part to be hoisted to the component that renders `TransactionColdstartRateChart`. I think having the data fetching _in_ the component kind of defeats the purpose of having two separate endpoints. It's kind of annoying that `TransactionCharts` is in between and I can see how it would be a big change. I'm OK with a follow-up ticket

gbamparop replied:
I agree , created this follow-up https://github.com/elastic/kibana/issues/124106

dgieselaar commented:
nice job @gbamparop and @AlexanderWert!

formgeist commented:
> ![image](https://user-images.githubusercontent.com/4104278/151768976-b1a28a94-ca85-4eaa-a306-31b78e869977.png)

@gbamparop Do you think it's possible that when we list out badges that just keep the badges to float as exemplified here;

<img width="353" alt="CleanShot 2022-01-31 at 10 28 09@2x" src="https://user-images.githubusercontent.com/4104278/151769123-2c89006e-d2bf-4f57-9d46-2e1aa59e0db5.png">

I imagine that list could be quite large depending on how many functions you include in a given service?

gbamparop commented:
> @gbamparop Do you think it's possible that when we list out badges that just keep the badges to float as exemplified here;

As discussed we decided to go with the current version for now as there can be several functions with large names and the popup can get quite large.

gbamparop committed:
Fix build errors

kibanamachine commented:
## 💔 Backport failed
The pull request could not be backported due to the following error:
`There are no branches to backport to. Aborting.`

### How to fix
Re-run the backport manually:
```
node scripts/backport --pr 122775
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
